rfcs/001-deterministic-states-and-caching.md

```diff
@@ -0,0 +1,716 @@
+
+  if (!data)
+    return (
+      <button disabled={loading} onClick={async () => await signMessage()}>
+        Sign message
+      </button>
+    )
+
+  return (
+    <div>
+      {data && <div>Signature: {data}</div>}
+      {error && <div>Error signing message</div>}
+    </div>
+  )
+}
+```
+
+This proposal adds a set of states (`isIdle`, `isLoading`, `isSuccess` & `isError`) that represents a finite state machine (only one flag can be truthy at the one time), which makes the following render code easier to read & follow:
+
+```tsx
+import { useSignMessage } from 'wagmi'
+
+const App = () => {
+  const [{ data, error, isIdle, isLoading, isSuccess, isError }, signMessage] = useSignMessage({
+    message: 'gm wagmi frens',
+  })
+
+  return (
+    <div>
+      {isIdle && (
+        <button disabled={loading} onClick={async () => await signMessage()}>
+          Sign message
+        </button>
+      )}
+      {isLoading && <div>Signing...</div>}
+      {isSuccess && <div>Signature: {data}</div>}
+      {isError && <div>Error signing message</div>}
+    </div>
+  )
+}
+```
+
+## Example 2 (caching)
+
+Some hooks in this library could be cacheable, such as `useAccount`, `useBalance`, `useNetwork`, etc. 
+
+There are no examples for caching, as it will be performed under-the-hood. 
+
+However, users' should be able to do either of the following:
+
+- change the cache expiry (time the data remains in the cache, cached results are shown to the user straight away),
+
+```tsx
+const [...] = useAccount({ cacheTime: 60000 });
+```
+
+- change the stale expiry (time when the data is considered stale, and then should perform a background refresh),
+
+```tsx
+const [...] = useAccount({ staleTime: 10000 });
+```
+
+- disable the cache completely.
+
+```tsx
+const [...] = useAccount({ cacheTime: 0 });
+```
+
+# Motivation
+
+The motivation of this RFC can be seen in the below discussion:
+
+https://github.com/tmm/wagmi/discussions/85
+
+TLDR; 
+
+- Non-deterministic & implicit states can lead to "impossible" states, which can cause UX bugs quite easily,
+- No idle state & implicit states lead to a confusing DX,
+- Most dapps experience a "flash of loading wallet/resolving ens" experience.
+
+# Detailed design
+
+Below are some (opinionated) notes that we should consider:
+
+## A note on hook categorization
+
+For the purpose of clarity, I think we should introduce the terminology of "read hooks" and "action hooks" or something similar.
+
+This will allow the consumer to know which hooks will invoke on mount (read hooks), and the ones whos actions won't be invoked on mount (action hooks).
+
+## A note on hook naming
+
+I feel like the naming of read hooks such as `useAccount`, `useBalance`, `useNetwork` are fine, however I reckon it would be nice to be more explicit with action hook naming to include intending verb (i.e. `useTransactionSend` instead of `useTransaction`). This also aligns with existing hooks such as `useContractWrite` and `useEnsLookup`.
```

Comment:
Agree about the naming change for `useTransaction` (and action hooks in general)